turbopack/crates/turbopack-core/src/issue/mod.rs

```diff
@@ -851,6 +852,18 @@ impl PlainSource {
     }
 }
 
+impl Ord for PlainSource {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.ident.cmp(&other.ident)
+    }
+}
+
+impl PartialOrd for PlainSource {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
```

Comment:
This violates the rules for `PartialOrd` in https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html:

![Screenshot 2024-11-13 at 2.43.07 PM.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/HAZVitxRNnZz8QMiPn4a/ba3d7c0a-1221-4e4b-9d96-bcff460db112.png)

`PartialOrd` must match the behavior of `PartialEq` (currently derived by the `turbo_tasks::value` macro). Not doing so can subtly break collection types like `BTreeSet` that rely on these rules.

Because you don't really expect to have two instances of `PlainSource` with the same `ident` but different values of `content` I think a reasonable thing to do would be to use pointer equality for `content`:

1. Expose an associated function on `ReadRef` to convert it into a pointer, with the same signature as [`std::sync::Arc::as_ptr`](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.as_ptr) (Note: [`triomphe::Arc::as_ptr`](https://docs.rs/triomphe/latest/triomphe/struct.Arc.html#method.as_ptr) wrongly uses a method, and not an associated function for this. [That violates rust's API guidelines](https://rust-lang.github.io/api-guidelines/predictability.html#c-smart-ptr). We should follow the stdlib's lead here.)
2. Compare the pointers here, and add a manual `PartialEq`/`Eq` implementation that uses `ptr_eq` on the `content` field.

Reply:
> add a manual PartialEq/Eq implementation that uses ptr_eq on the content field.

That would still violate the `PartialOrd` `PartialEq` thing, since PartialOrd would say Equal when the content ptr is not equal.

We also can't compare pointers, since that would persistent caching as pointers will change after deserialisation.
Also note that something that has equal pointers might have different pointers after deserialisation.

I will just do a full compare on the `content` for PartialOrd to create matching implementations

Reply:
I derived PartialOrd, it should be much safer and more correct now...